### PR TITLE
[BEAM-6877] trivial_inference: make remaining tests pass

### PR DESCRIPTION
**Pull request by udim**
- Partially support CALL_FUNCTION_EX (sans kwargs).
- Adds a few UNPACK opcodes (just enough to pass tests).
- A couple of tests use Py3 syntax, hidden behind an eval() (sorry).

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
cc: @tvalentyn

File: sdks/python/apache_beam/typehints/trivial_inference_test.py [hunk 1]
```diff
@@ -235,6 +235,36 @@ def m(self, x):
     self.assertReturnType(typehints.Any, lambda: A.m(A(), 3.0), depth=0)
     self.assertReturnType(float, lambda: A.m(A(), 3.0), depth=1)
 
+  # pylint: disable=eval-used
+  # TODO(udim): Remove eval() call once Python 2 support is removed from Beam.
+  @unittest.skipIf(sys.version_info < (3,), 'Python 3 only')
+  def testBuildListUnpack(self):
+    # Lambda uses BUILD_LIST_UNPACK opcode in Python 3.
+    # Uses eval() to hide Python 3 syntax from Python 2.
+    eval('''self.assertReturnType(typehints.List[int],
```
Comment (tvalentyn) [inline sdks/python/apache_beam/typehints/trivial_inference_test.py]: Nice trick, FYI, @Juta who ran into this problem in one of the Py3-only tests.

**Comment by tvalentyn** [comment]
Run Python PreCommit

**Comment by tvalentyn** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

**Comment by tvalentyn** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python precommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python precommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
R: @robertwb (in case you haven't seen all the emails)

**Comment by udim** [comment]
run python precommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
run python precommit

**Comment by udim** [comment]
run python postcommit

**Comment by udim** [comment]
Run Python 2 PostCommit

**Comment by tvalentyn** [comment]
Seed job did not run yet last time. Try again :-)

**Comment by udim** [comment]
Run Python 2 PostCommit

**Comment by udim** [comment]
Run Python 3.5 PostCommit

**Comment by udim** [comment]
Run Python 3.6 PostCommit

**Comment by udim** [comment]
Run Python 3.7 PostCommit

**Comment by udim** [comment]
Run Python 2 PostCommit

**Comment by udim** [comment]
Run Python 3.7 PostCommit

**Comment by aaltay** [comment]
@robertwb and @tvalentyn could you please review this?

Test failures at this point looks like unrelated flakes.

**Comment by udim** [comment]
Run Python_PVR_Flink PreCommit

**Comment by udim** [comment]
run python 2 postcommit

File: sdks/python/apache_beam/typehints/trivial_inference.py [hunk 2]
```diff
@@ -458,10 +458,18 @@ def infer_return_type_func(f, input_types, debug=False, depth=0):
...
-          pop_count = (arg & 1) + 3
-          return_type = Any
+          has_kwargs = arg & 1  # type: int
+          pop_count = has_kwargs + 1
+          if has_kwargs:
+            # TODO(udim): Unimplemented. Requires same functionality as a
+            #   CALL_FUNCTION_KW implementation.
+            return_type = Any
+          else:
+            args_and_callable = state.stack[-1]
```
Comment (tvalentyn) [inline sdks/python/apache_beam/typehints/trivial_inference.py]: I was reading https://docs.python.org/3/library/dis.html, which says:

> CALL_FUNCTION_EX(flags)
> Calls a callable object with variable set of positional and keyword arguments. If the lowest bit of flags is set, the top of the stack contains a mapping object containing additional keyword arguments. **Below that is an iterable object containing positional arguments and a callable object to call.** BUILD_MAP_UNPACK_WITH_CALL and BUILD_TUPLE_UNPACK_WITH_CALL can be used for merging multiple mapping objects and iterables containing arguments. Before the callable is called, the mapping object and iterable object are each “unpacked” and their contents passed in as keyword and positional arguments respectively. CALL_FUNCTION_EX pops all arguments and the callable object off the stack, calls the callable object with those arguments, and pushes the return value returned by the callable object.
> 

I was not sure whether the phrase in bold means that callable and arguments are stored in the one element of the stack `state.stack[-1]`,  or arg list and callable  are stored in two consecutive elements of the stack below, i.e. arg in state.stack[-1], and callable in state.stack[-2].

I wrote this snippet:
```
from apache_beam.typehints import trivial_inference
def f(*kwargs):
    return kwargs[0] + kwargs[1]

def g():
    return f(*[5, 6])

def run(argv=None):
  print(g())
  trivial_inference.infer_return_type_func(g, [], debug=True)

if __name__ == '__main__':
    run()
```
When I run this code on this PR, I get an error, that does not happen on master. I used Python 3.6. 

(py3) :python$ python -m test

```
/home/valentyn/projects/beam/beam/beam/sdks/python/apache_beam/__init__.py:84: UserWarning: Some syntactic constructs of Python 3 are not yet fully supported by Apache Beam.
  'Some syntactic constructs of Python 3 are not yet fully supported by '
11

#####
<function g at 0x7f8a2d7ac9d8> 140231445236184 []
  6           0 LOAD_GLOBAL              0 (f)
              2 LOAD_CONST               1 (5)
              4 LOAD_CONST               2 (6)
              6 BUILD_LIST               2
              8 CALL_FUNCTION_EX         0
             10 RETURN_VALUE
        0 LOAD_GLOBAL              0 (f) Executing simple op LOAD_GLOBAL

Stack: [Const[<function f at 0x7f8a2d7ac950>]] Vars: []
{}
        2 LOAD_CONST               1 (5) Executing simple op LOAD_CONST

Stack: [Const[<function f at 0x7f8a2d7ac950>], Const[5]] Vars: []
{}
        4 LOAD_CONST               2 (6) Executing simple op LOAD_CONST

Stack: [Const[<function f at 0x7f8a2d7ac950>], Const[5], Const[6]] Vars: []
{}
        6 BUILD_LIST               2 Executing simple op BUILD_LIST

Stack: [Const[<function f at 0x7f8a2d7ac950>], List[int]] Vars: []
{}
        8 CALL_FUNCTION_EX         0 has kwargs: 0
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/valentyn/projects/beam/beam/beam/sdks/python/test.py", line 14, in <module>
    run()
  File "/home/valentyn/projects/beam/beam/beam/sdks/python/test.py", line 10, in run
    trivial_inference.infer_return_type_func(g, [], debug=True)
  File "/home/valentyn/projects/beam/beam/beam/sdks/python/apache_beam/typehints/trivial_inference.py", line 471, in infer_return_type_func
    return_type = infer_return_type(args_and_callable[-1].value,
TypeError: 'ListConstraint' object does not support indexing
```
What do you think, @udim?

Comment (udim) [reply sdks/python/apache_beam/typehints/trivial_inference.py]: I think you're right.
The only test cases I had so far had BUILD_TUPLE_UNPACK_WITH_CALL followed by CALL_FUNCTION_EX. I've added an additional test case.

**Comment by udim** [comment]
run python 2 postcommit

File: sdks/python/apache_beam/typehints/trivial_inference.py [hunk 3]
```diff
@@ -458,10 +458,27 @@ def infer_return_type_func(f, input_types, debug=False, depth=0):
           pop_count = arg + 2
           return_type = Any
         elif opname == 'CALL_FUNCTION_EX':
-          # TODO(udim): Handle variable argument lists. Requires handling kwargs
-          #   first.
-          pop_count = (arg & 1) + 3
```
Comment (tvalentyn) [inline sdks/python/apache_beam/typehints/trivial_inference.py]: In new version we increment pop_count by has_kwargs + 2, do you think old version just had a typo?  
```
          pop_count = (arg & 1) + 3	
```

File: sdks/python/apache_beam/typehints/trivial_inference_test.py [hunk 4]
```diff
@@ -235,6 +235,60 @@ def m(self, x):
...
+                          lambda x1, x2, _list: fn(x1, x2, *_list),
+                          [str, float, typehints.List[int]])
+    # No *args
+    self.assertReturnType(typehints.Tuple[str, typehints.List[int]],
+                          lambda x1, x2, _list: fn(x1, x2, *_list),
+                          [str, typehints.List[int]])
+
+  @unittest.skipIf(sys.version_info < (3, 6), 'CALL_FUNCTION_EX is new in 3.6')
+  def testCallFunctionEx(self):
+    # Test when fn arguments are built using BUiLD_LIST.
```
Comment (tvalentyn) [inline sdks/python/apache_beam/typehints/trivial_inference_test.py]: nit: BUILD_LIST

**Comment by tvalentyn** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
Run Python 2 PostCommit

**Comment by aaltay** [comment]
Run Python 2 PostCommit

File: sdks/python/apache_beam/typehints/trivial_inference.py (hunk 3 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/trivial_inference.py]: Not a typo but a bug. In python `stack[-pop_count:]` doesn't trigger an error even though `stack[-pop_count]` is out of bounds.

**Comment by udim** [comment]
run python 3.5 postcommit

File: sdks/python/apache_beam/typehints/trivial_inference_test.py (hunk 4 above)
Comment (udim) [reply sdks/python/apache_beam/typehints/trivial_inference_test.py]: I'll leave this for my next PR as I don't want to reset the tests.

**Comment by aaltay** [comment]
All tests are failing with the same 404 not found error related to BEAM-6202. https://github.com/apache/beam/pull/9182 would fix that error.

**Comment by udim** [comment]
run python 3.6 postcommit

**Comment by udim** [comment]
run python 3.7 postcommit

**Comment by aaltay** [comment]
One more fix related to retrying on 404 errors: https://github.com/apache/beam/pull/9196

**Comment by udim** [comment]
Thanks Ahmet. I'm satisfied with the tests here - merging.

**Pull request merged**